File: cgrcparser.cpp
/*
 * This file is part of cgrc.
 *
 * Copyright (c) 2022 Luca Carlon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Author:  Luca Carlon
 * Date:    2022.08.06
 * Company: -
 */

#include <QCoreApplication>
#include <QCommandLineParser>
#include <QTextStream>
#include <QDebug>

#include "cgrcparser.h"
#include "cgrc_data.h"
#include "cgrc_conf.h"

#include "lqtutils/lqtutils_string.h"

const QString CGRCParser::RESET_SEQUENCE = QSL("\x1b[0m");

CGRCParser::CGRCParser()
{}

QString CGRCParser::parseLogLine(const QList<CGRCConfItem>& confItems, const QString& inLine)
{
#if QT_VERSION >= QT_VERSION_CHECK(5, 15, 0)
    qsizetype inLineLength = inLine.length();
#else
    int inLineLength = inLine.length();
#endif
    QString outLine = inLine;
    const CGRCColorItem** charColors = new const CGRCColorItem*[inLineLength];
    memset(charColors, 0, sizeof(CGRCColorItem*)*inLineLength);

#if DEBUG_LOG
    qInfo() << "-> Processing:" << inLine;
#endif

    bool stopProcessing = false;
    for (const CGRCConfItem& confItem : confItems) {
        QRegularExpressionMatchIterator matches = confItem.regexp.globalMatch(inLine);
        if (stopProcessing)
            break;
        if (matches.hasNext() && confItem.skip)
            return QString();
        stopProcessing = (confItem.countMode == CGRC_COUNT_STOP && matches.hasNext());
        while (matches.hasNext()) {
#if DEBUG_LOG
            qDebug() << "Regex matches:" << confItem.regexp.pattern();
#endif
            QRegularExpressionMatch match = matches.next();
            for (int i = 0; i < match.capturedLength() && i < confItem.colors.length(); i++) {
                int from = match.capturedStart(i);
                int to   = match.capturedEnd(i);
                for (int j = from; j < to; j++) {
                    if (!confItem.colors[i].attrs().contains(CGRC_NONE))
                        charColors[j] = &(confItem.colors[i]);
                }
            }

            if (confItem.countMode == CGRC_COUNT_ONCE) {
                break;
            }
        }
    }

    QString formattedLine;
    const CGRCColorItem* lastColor = charColors[0];
    int lastIndex = 0;
    for (int i = 1; i < inLineLength + 1; i++) {
        if (charColors[i] == lastColor && i != inLineLength)
            continue;
        int from = lastIndex;
        int to   = i;
        QString formattedSequence;
        if (!lastColor)
            formattedSequence = RESET_SEQUENCE + inLine.mid(lastIndex, i - lastIndex);
        else
            formattedSequence = lastColor->escapeSequence() + (inLine.mid(lastIndex, i - lastIndex)) + lastColor->clearSequence();

#if DEBUG_LOG
        qDebug() << inLine.mid(lastIndex, i - lastIndex) << "->" << formattedSequence;
#endif

        formattedLine.append(formattedSequence);

        lastIndex = i;
        lastColor = charColors[i];
    }

    formattedLine.append(RESET_SEQUENCE);

    delete[] charColors;

    return formattedLine;
}

void CGRCParser::parseCmd(const QCoreApplication* app, QCommandLineParser* parser)
{
    QCommandLineOption optHelp(QSL("version"),
                               QSL("Shows the version of the application."));
    QCommandLineOption optListLocations(QSL("list-locations"),
                                        QSL("Shows the location to use for the conf files."));
    QCommandLineOption optListConfs(QSL("list-configurations"),
                                    QSL("List available configurations."));
    QCommandLineOption optConfFilePath(QSL("conf-path"),
                                       QSL("The configuration argument is to be interpreted as a path to the configuration file."));
    QCommandLineOption optInputFilePath(QSL("input-path"),
                                        QSL("Path to input file."),
                                        QSL("input-path"));

#define APP_DESCRIPTION                                               \
    "Configurable terminal text formatter\n\n"                        \
    "cgrc formats text from stdin according to custom configuration " \
    "files and outputs the result with ANSI escape codes to stdout. " \
    "Configuration files includes a set of regular expressions with " \
    "the related format to be used to the match and the captures."

    parser->setApplicationDescription(QSL(APP_DESCRIPTION));
    parser->addHelpOption();
    parser->addPositionalArgument("conf", "Configuration file");
    parser->addOption(optHelp);
    parser->addOption(optListLocations);
    parser->addOption(optListConfs);
    parser->addOption(optConfFilePath);
    parser->addOption(optInputFilePath);
    parser->process(app->arguments());
}

QList<CGRCColorItem> CGRCParser::parseColors(const QString& colors)
{
    QList<CGRCColorItem> items;
    const QStringList captures = colors.split(QSL(","));
    for (const QString& capture : captures) {

#if QT_VERSION >= QT_VERSION_CHECK(5, 15, 0)
        const QStringList options = capture.split(QSL(" "), Qt::SkipEmptyParts);
#else
        const QStringList options = capture.split(QSL(" "), QString::SkipEmptyParts);
#endif

        QSet<CGRC_Attrib> attrs;
        LC_LogColor forg = LC_LogColor::LC_FORG_COL_DEFAULT;
        LC_BackColor back = LC_BackColor::LC_BACK_COL_DEFAULT;
        for (const QString& option : options) {
            // Attributes.
            QString lowerOption = option.toLower();
            if (COLORS_ATTRS.contains(lowerOption))
                attrs.insert(COLORS_ATTRS.value(lowerOption));
            if (COLORS_BACK.contains(lowerOption))
                back = COLORS_BACK.value(lowerOption);
            if (COLORS_FORG.contains(lowerOption))
                forg = COLORS_FORG.value(lowerOption);
        }

        CGRCColorItem colorItem(attrs, forg, back);
        items.append(colorItem);
    }

    return items;
}

CGRCConf CGRCParser::parseConf(QFile& confFile)
{
    CGRCConf ret;
    QList<CGRCConfItem>& retItems = ret.items;
    CGRCConfItem currentItem;
    QTextStream stream(&confFile);
    while (!stream.atEnd()) {
        QString line = stream.readLine();
        if (line.startsWith(QSL("desc="))) {
            ret.description = line.replace(QSL("desc="), QSL(""));
            continue;
        }

        if (line.startsWith(QSL("regexp="))) {
            currentItem.regexp = QRegularExpression(line.replace(QSL("regexp="), QSL("")));
            if (!currentItem.regexp.isValid()) {
                QString errorString = currentItem.regexp.errorString();
                int errorOffset = currentItem.regexp.patternErrorOffset();
                qWarning() << "Invalid regexp found in line:"
                           << line
                           << "->"
                           << errorString
                           << "at"
                           << errorOffset;
            }
            continue;
        }

        if (line.startsWith(QSL("colours="))) {
            currentItem.colors.append(parseColors(line.replace(QSL("colours="), QSL(""))));
            continue;
        }

        if (line.startsWith(QSL("skip="))) {
            currentItem.skip = (line.compare(QSL("skip=yes"), Qt::CaseInsensitive) == 0);
            continue;
        }

        if (line.startsWith(QSL("count="))) {
            if (line.compare(QSL("count=once"), Qt::CaseInsensitive) == 0) {
                currentItem.countMode = CGRC_COUNT_ONCE;
                continue;
            }
            else if (line.compare(QSL("count=more"), Qt::CaseInsensitive) == 0) {
                currentItem.countMode = CGRC_COUNT_MORE;
                continue;
            }
            else if (line.compare(QSL("count=stop"), Qt::CaseInsensitive) == 0) {
                currentItem.countMode = CGRC_COUNT_STOP;
                continue;
            }
            else if (line.compare(QSL("count=previous"), Qt::CaseInsensitive) == 0) {
                currentItem.countMode = CGRC_COUNT_PREVIOUS;
                continue;
            }
            else if (line.compare(QSL("count=block"), Qt::CaseInsensitive) == 0) {
                currentItem.countMode = CGRC_COUNT_BLOCK;
                continue;
            }
            else if (line.compare(QSL("count=unblock"), Qt::CaseInsensitive) == 0) {
                currentItem.countMode = CGRC_COUNT_UNBLOCK;
                continue;
            }
            else
                qWarning() << "Invalid count mode";
            break;
        }

        if (!currentItem.regexp.pattern().isEmpty()) {
            retItems.append(currentItem);
            currentItem.clear();
        }
    }

    if (!retItems.contains(currentItem))
        retItems.append(currentItem);

    return ret;
}

File: cgrc-rust/src/cgrcparser.rs
/*
 * This file is part of cgrc.
 *
 * Copyright (c) 2023 Luca Carlon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

use std::{process, fs::File, io::{BufReader, BufRead, Cursor}, collections::HashSet};
use std::ptr;
use fancy_regex::Regex;
use crate::cgrcdata::{
    CgrcColorItem,
    CgrcCountMode,
    CgrcConfItem,
    CgrcConf,
    CgrcAttrib,
    LcLogColor,
    LcBackColor,
    COLORS_ATTRS,
    COLORS_FORG,
    COLORS_BACK
};

pub struct CGRCParser {}

impl CGRCParser {
    ///
    /// Parsers a conf file.
    /// 
    pub fn parse_conf(conf_file: &String) -> CgrcConf {
        let file = match File::open(conf_file) {
            Err(e) => {
                log::error!("Failed to open conf file {conf_file}: {e}");
                process::exit(1);
            },
            Ok(f) => {
                f
            }
        };

        let reader = BufReader::new(file);
        return Self::parse_conf_lines(reader);
    }

    ///
    /// Parse lines from a buffered reader.
    /// 
    pub fn parse_conf_lines<T: BufRead>(reader: T) -> CgrcConf {
        let mut item = CgrcConfItem::new();
        let mut conf = CgrcConf {
            description: None,
            items: vec![]
        };
        for line in reader.lines() {
            match line {
                Err(e) => {
                    log::error!("Failed to read conf file: {e}");
                    process::exit(1);
                },
                Ok(line) => {
                    if CGRCParser::parse_conf_line(&line, &mut conf,  &mut item) {
                        conf.items.push(item.clone());
                        item = CgrcConfItem::new();
                    }
                }
            }
        }

        if item.regex.is_some() {
            conf.items.push(item.clone());
        }

        conf
    }

    ///
    /// Parses a configuration from a string.
    /// 
    pub fn parse_conf_string(conf: String) -> CgrcConf {
        let cursor = Cursor::new(conf);
        let reader = BufReader::new(cursor);
        return CGRCParser::parse_conf_lines(reader);
    }

    // Private portion
    // ===============
    fn parse_conf_line(line: &String, conf: &mut CgrcConf, item: &mut CgrcConfItem) -> bool {
        let lline = line.to_lowercase();

        if lline.starts_with("desc=") {
            conf.description = Some(lline.replace("desc=", ""));
            return false;
        }

        if lline.starts_with("regexp=") {
            item.regex = match Regex::new(format!("{}", line.replace("regexp=", "")).as_str()) {
                Err(e) => {
                    log::error!("Failed to parse regex: {line}");
                    log::error!("{}", e.to_string());
                    process::exit(1);
                },
                Ok(r) => Some(r)
            };
            return false;
        }

        if lline.starts_with("colours=") {
            item.colors.append(&mut CGRCParser::parse_colors(&lline.replace("colours=", "")));
            return false;
        }

        if lline.starts_with("skip=") {
            item.skip = Some(lline.to_lowercase() == "skip=yes");
            return false;
        }

        if lline.starts_with("count=") {
            match lline.as_str() {
                "count=once" => {
                    item.count_mode = Some(CgrcCountMode::CgrcCountOnce);
                },
                "count=more" => {
                    item.count_mode = Some(CgrcCountMode::CgrcCountMore);
                },
                "count=stop" => {
                    item.count_mode = Some(CgrcCountMode::CgrcCountStop)
                },
                "count=previous" => {
                    item.count_mode = Some(CgrcCountMode::CgrcCountPrevious)
                },
                "count=block" => {
                    item.count_mode = Some(CgrcCountMode::CgrcCountBlock)
                },
                "count=unblock" => {
                    item.count_mode = Some(CgrcCountMode::CgrcCountUnblock)
                }
                _default => {
                    log::error!("Invalid count mode");
                    process::exit(1);
                }
            }
            return false;
        }

        return item.regex.is_some();
    }

    fn parse_colors(line: &String) -> Vec<CgrcColorItem> {
        let mut items: Vec<CgrcColorItem> = vec![];
        let line_tokens = line.split(",");
        for line_token in line_tokens {
            let options = line_token.split(" ");
            let mut attrs: HashSet<CgrcAttrib> = HashSet::new();
            let mut forg = LcLogColor::LcForgColDefault;
            let mut back: LcBackColor = LcBackColor::LcBackColDefault;
            for option in options {
                let lower_option = option.to_lowercase();
                if COLORS_ATTRS.contains_key(option) {
                    attrs.insert(COLORS_ATTRS.get(&lower_option).unwrap().clone());
                    continue;
                }
                if COLORS_BACK.contains_key(option) {
                    back = COLORS_BACK.get(&lower_option).unwrap().clone();
                    log::info!("Found color: {:?}", back);
                    continue;
                }
                if COLORS_FORG.contains_key(option) {
                    forg = COLORS_FORG.get(&lower_option).unwrap().clone();
                    continue;
                }
            }

            let item = CgrcColorItem::new(attrs, forg, back);
            items.push(item);
        }

        items
    }

    ///
    /// Parses the line.
    /// 
    pub fn parse_log_line(conf_items: &Vec<CgrcConfItem>, in_line: &String, debug: bool) -> Option<String> {
        let in_line_length = in_line.len();
        let mut char_colors: Vec<*const CgrcColorItem> = vec![ptr::null(); in_line_length];
        let mut stop_processing = false;
        for conf_item in conf_items {
            if debug {
                log::debug!("Testing conf: {:?}", conf_item);
            }
            
            if stop_processing {
                break;
            }

            let count_mode = conf_item.count_mode.as_ref().unwrap_or(&CgrcCountMode::CgrcCountMore);
            let regex = conf_item.regex.as_ref().unwrap();
            for regex_match in regex.captures_iter(&in_line) {
                if let Ok(regex_match) = regex_match {
                    if conf_item.skip.unwrap_or(false) {
                        return None;
                    }
                    
                    for i in 0..regex_match.len() {
                        if i >= conf_item.colors.len() {
                            break;
                        }
    
                        let capture = match regex_match.get(i) {
                            None => continue,
                            Some(v) => v
                        };
                        if debug {
                            log::debug!("Captured: {:?}", capture.as_str());
                        }
                        let from = capture.start();
                        let to   = capture.end();
                        for j in from..to {
                            if !conf_item.colors[i].attrs.contains(&CgrcAttrib::CgrcNone) {
                                char_colors[j] = &(conf_item.colors[i]);
                                if debug {
                                    log::warn!("Color: {:?}", conf_item.colors[i]);
                                }
                            }
                        }
                    }

                    stop_processing = count_mode == &CgrcCountMode::CgrcCountStop;

                    if count_mode == &CgrcCountMode::CgrcCountOnce {
                        break;
                    }
                }
            }
            
        }

        let mut formatted_line = String::new();
        let mut formatted_seq;
        let mut last_color = char_colors[0];
        let mut last_index = 0;
        for i in 1..=in_line_length {
            if i != in_line_length && char_colors[i] == last_color {
                continue;
            }
            if last_color.is_null() {
                formatted_seq = "\x1b[0m".to_string().to_string() + &in_line[last_index..i].to_owned();
            }
            else {
                unsafe {
                    formatted_seq = (*last_color).escape_seq.to_owned()
                        + &in_line[last_index..i].to_owned()
                        + "\x1b[0m".to_string().as_str();
                }
            }

            formatted_line += &formatted_seq;

            if i != in_line_length {
                last_index = i;
                last_color = char_colors[i];
            }
        }

        formatted_line += &"\x1b[0;0m".to_string();

        return Some(formatted_line);
    }
}
